tls-tutorial/doc/tls_in_udfs.md

```diff
@@ -0,0 +1,225 @@
+
+You will also need:
+
+1. A Java Development Kit (JDK 11 or later, we recommend [OpenJDK](https://openjdk.org/install/))
+2. `keytool` program (comes with the JDK)
+3. [`curl`](https://curl.se/) to access BucketFS and download file from the internet
+
+## UDFs
+
+A User Defined Function (short "UDF") is a way to extend the Exasol database with new features that can be called from within Exasol SQL statements. Basically. A UDF is a program that runs on the Exasol cluster in the context of an SQL query. The UDF gets its parameters from the SQL statement and its return values can be used by other parts of the query.
+
+Each UDF runs in a sandbox that only lives in the context of a single statement. This ensures that UDF can't be used to escape the boundaries of the statement it belongs too and prevents data leakage.
+
+While the sandbox makes the UDF more secure, it unfortunately makes using TLS harder.
+
+## Language Containers
+
+A UDF can be written in an arbitrary programming language but requires an appropriate runtime environment that can be provided as a so-called "Language Container". The Exasol database ships with a standard container providing the runtime environment for Java and Python. Other containers and updates for the standard container are [available for download on GitHub](https://github.com/exasol/script-languages-release/releases).
+
+All UDFs require such a language container &mdash; except Lua UDFs, since Lua is embedded directly into the Exasol engine.
+
+## Running UDFs
+
+Whenever you run a SQL statement that contains a UDF, Exasol starts the corresponding language container. This serves both as a runtime environment and as a sandbox at the same time. Everything the UDF does happens inside the running container. Once the statement's execution is complete, Exasol terminates the container.
+
+## Certificate Validation in UDFs
+
+The introduction talked about how certificates are used to authenticate communication partners in TLS. As you might remember, you need a valid chain of certificates down to a certification agency (CA) that you trust.
+
+Since the UDF runs in a sandbox, it can only see what the sandbox provides. That means without extra effort the UDF can only validate TLS certificates down to a CA certificate available in the sandbox (i.e. the language container). This is usually no problem if you want to access a service on the Internet, because in almost all cases services exposed to the internet have certificates that can be traced down to one of the well-established CAs. And the certificates for these CAs are shipped with the language container.
+
+### Listing the Standard Certificates
+
+This tutorial comes with a UDF that can be built into the JAR `tls-tutorial.jar`. You can install the JAR by following these steps:
+
+1. Change to the base directory of this `exasol-java-tutorial`
+2. If you haven't already, build the binaries (JAR files)
+   ```shell
+   mvn clean verify
```

Comment:
Will we provide a pre-complied JAR as well? Say, as a release.

Reply:
Yes. CI build will be another PR.